File: mekong-cli/lib/sdk.js
'use strict'
/**
 * mekong-cli/sdk — programmatic API for Node.js projects.
 *
 * Usage:
 *   const mekong = require('mekong-cli/sdk')
 *
 *   // Expose a local port (returns public URL)
 *   const { url, stop } = await mekong.expose(3000)
 *   console.log('Public:', url)
 *   // …later…
 *   stop()
 *
 *   // Auth helpers
 *   const token = mekong.getToken()
 *   await mekong.login()   // opens browser, waits for approval
 *   mekong.logout()        // removes ~/.mekong/config.json
 */

const { spawn }   = require('child_process')
const { findMekong } = require('./find-mekong')
const os   = require('os')
const fs   = require('fs')
const path = require('path')
const http = require('https')

const API_BASE = 'https://api.angkorsearch.dev'
const WEB_BASE = 'https://angkorsearch.dev'

// ── Config helpers ────────────────────────────────────────────────────────────

function mekongDir () {
  return path.join(os.homedir(), '.mekong')
}

function configPath () {
  return path.join(mekongDir(), 'config.json')
}

/**
 * Return the saved API token, or null if not logged in.
 * Also checks MEKONG_TOKEN env var.
 */
function getToken () {
  if (process.env.MEKONG_TOKEN) return process.env.MEKONG_TOKEN
  try {
    const cfg = JSON.parse(fs.readFileSync(configPath(), 'utf8'))
    return cfg.token || null
  } catch {
    return null
  }
}

/** Return saved auth config or null. */
function getAuthConfig () {
  try {
    return JSON.parse(fs.readFileSync(configPath(), 'utf8'))
  } catch {
    return null
  }
}

// ── HTTP helpers ──────────────────────────────────────────────────────────────

function httpsGet (url) {
  return new Promise((resolve, reject) => {
    http.get(url, res => {
      let body = ''
      res.on('data', chunk => body += chunk)
      res.on('end', () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(body) }) }
        catch { resolve({ status: res.statusCode, body: {} }) }
      })
    }).on('error', reject)
  })
}

function httpsPost (url) {
  return new Promise((resolve, reject) => {
    const u = new URL(url)
    const options = {
      hostname: u.hostname,
      path: u.pathname + u.search,
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Content-Length': 0 },
    }
    const req = http.request(options, res => {
      let body = ''
      res.on('data', chunk => body += chunk)
      res.on('end', () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(body) }) }
        catch { resolve({ status: res.statusCode, body: {} }) }
      })
    })
    req.on('error', reject)
    req.end()
  })
}

function unwrap (body) {
  return body?.data ?? body
}

// ── expose() ─────────────────────────────────────────────────────────────────

/**
 * Expose a local port via the mekong binary.
 * Returns a Promise<{ url: string, stop: () => void }>.
 *
 * @param {number} port  - local port to expose
 * @param {object} [opts]
 * @param {string}  [opts.token]    - API token (falls back to saved login / MEKONG_TOKEN)
 * @param {string}  [opts.expire]   - tunnel lifetime e.g. "2h", "1d"
 * @param {boolean} [opts.daemon]   - run in background (-d)
 * @param {boolean} [opts.noQr]     - suppress QR code
 * @param {string}  [opts.binary]   - custom path to mekong binary
 * @returns {Promise<{ url: string, stop: () => void }>}
 */
function expose (port, opts = {}) {
  return new Promise((resolve, reject) => {
    const binary = opts.binary || findMekong()
    if (!binary) {
      return reject(new Error(
        'mekong binary not found. Install from https://github.com/MuyleangIng/MekongTunnel/releases'
      ))
    }

    const token = opts.token || getToken()

    const args = [String(port)]
    if (opts.expire)         args.push('--expire', opts.expire)
    if (token)               args.push('--token', token)
    if (opts.daemon)         args.push('-d')
    if (opts.noQr !== false) args.push('--no-qr') // default true in SDK mode

    const proc = spawn(binary, args, { stdio: ['ignore', 'pipe', 'pipe'] })

    const ansiRe = /\x1b\[[0-9;]*[A-Za-z]/g
    const urlRe  = /https?:\/\/[a-zA-Z0-9][a-zA-Z0-9\-.\/]*[a-zA-Z0-9]/
    let resolved = false

    function handleChunk (chunk) {
      const clean = chunk.toString().replace(ansiRe, '')
      if (!resolved) {
        const m = clean.match(urlRe)
        if (m) {
          resolved = true
          resolve({ url: m[0], stop: () => proc.kill('SIGTERM') })
        }
      }
    }

    proc.stdout.on('data', handleChunk)
    proc.stderr.on('data', handleChunk)

    proc.on('exit', code => {
      if (!resolved) {
        reject(new Error(`mekong exited with code ${code} before tunnel was ready`))
      }
    })

    // Timeout after 30 s if no URL received
    setTimeout(() => {
      if (!resolved) {
        proc.kill('SIGTERM')
        reject(new Error('Timed out waiting for tunnel URL'))
      }
    }, 30_000)
  })
}

// ── login() ──────────────────────────────────────────────────────────────────

/**
 * Open browser → mekongtunnel.dev login page → poll until token received.
 * Saves token to ~/.mekong/config.json.
 * Returns the token string.
 *
 * @returns {Promise<string>} resolved token
 */
async function login () {
  // 1. Create device session
  const { body: raw, status } = await httpsPost(API_BASE + '/api/cli/device')
  if (status !== 200) throw new Error(`Server returned ${status}`)
  const sess = unwrap(raw)
  if (!sess?.session_id) throw new Error('Unexpected server response')

  console.log('\n  Open this URL to log in:')
  console.log('  \x1b[35m' + sess.login_url + '\x1b[0m\n')

  // Try to open browser
  const opener = process.platform === 'win32' ? 'start'
    : process.platform === 'darwin' ? 'open' : 'xdg-open'
  try {
    const { execSync } = require('child_process')
    execSync(`${opener} "${sess.login_url}"`, { stdio: 'ignore' })
  } catch {}

  console.log('  Waiting for authorization', { end: '' })

  // 2. Poll
  const deadline = Date.now() + 5 * 60 * 1000
  const interval = (sess.poll_interval || 3) * 1000

  while (Date.now() < deadline) {
    await new Promise(r => setTimeout(r, interval))
    const { body: pb } = await httpsGet(
      API_BASE + '/api/cli/device?session_id=' + sess.session_id
    )
    const poll = unwrap(pb)
    if (poll?.status === 'approved' && poll.token) {
      // Save to disk
      fs.mkdirSync(mekongDir(), { recursive: true })
      fs.writeFileSync(configPath(), JSON.stringify({ token: poll.token }, null, 2), { mode: 0o600 })
      console.log('\n  \x1b[32m✔  Logged in!\x1b[0m\n')
      return poll.token
    }
    if (poll?.status === 'expired') {
      throw new Error('Session expired — call login() again')
    }
    process.stdout.write('.')
  }
  throw new Error('Login timed out')
}

// ── logout() ─────────────────────────────────────────────────────────────────

/**
 * Remove saved credentials from ~/.mekong/config.json.
 */
function logout () {
  try {
    fs.unlinkSync(configPath())
  } catch {}
}

// ── whoami() ─────────────────────────────────────────────────────────────────

/**
 * Return saved auth config { token, email, user_id } or null.
 */
function whoami () {
  return getAuthConfig()
}

module.exports = { expose, login, logout, whoami, getToken }

File: mekong-python-sdk/src/mekong_tunnel/sdk.py
"""
mekong_tunnel.sdk — Programmatic Python API for MekongTunnel.

Usage::

    import mekong_tunnel as mekong

    # Expose a local port (blocking until tunnel URL is received)
    tunnel = mekong.expose(8000)
    print("Public URL:", tunnel.url)
    # …later…
    tunnel.stop()

    # Auth helpers
    token = mekong.get_token()
    mekong.login()   # opens browser, waits for approval, saves ~/.mekong/config.json
    mekong.logout()  # removes saved credentials
    info = mekong.whoami()
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
import threading
import time
import urllib.request
import urllib.error
import webbrowser
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

API_BASE = "https://api.angkorsearch.dev"
WEB_BASE = "https://angkorsearch.dev"

# ── Config helpers ────────────────────────────────────────────────────────────

def _mekong_dir() -> Path:
    return Path.home() / ".mekong"


def _config_path() -> Path:
    return _mekong_dir() / "config.json"


def get_token() -> Optional[str]:
    """Return the API token from env var or saved login, or None."""
    if tok := os.environ.get("MEKONG_TOKEN", "").strip():
        return tok
    try:
        cfg = json.loads(_config_path().read_text())
        return cfg.get("token") or None
    except Exception:
        return None


def _load_config() -> Optional[dict]:
    try:
        return json.loads(_config_path().read_text())
    except Exception:
        return None


def _save_config(cfg: dict) -> None:
    _mekong_dir().mkdir(parents=True, exist_ok=True)
    p = _config_path()
    p.write_text(json.dumps(cfg, indent=2))
    p.chmod(0o600)


# ── HTTP helpers ──────────────────────────────────────────────────────────────

def _api_post(path: str) -> dict:
    url = API_BASE + path
    req = urllib.request.Request(url, data=b"", method="POST",
                                 headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.loads(resp.read())


def _api_get(path: str) -> dict:
    with urllib.request.urlopen(API_BASE + path, timeout=15) as resp:
        return json.loads(resp.read())


def _unwrap(body: dict):
    return body.get("data", body)


# ── Tunnel handle ─────────────────────────────────────────────────────────────

@dataclass
class Tunnel:
    """Handle to a running tunnel process."""
    url: str
    port: int
    _proc: subprocess.Popen = field(repr=False)

    def stop(self) -> None:
        """Terminate the tunnel process."""
        try:
            self._proc.terminate()
        except Exception:
            pass

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.stop()

    def __repr__(self):
        return f"Tunnel(url={self.url!r}, port={self.port})"


# ── expose() ─────────────────────────────────────────────────────────────────

def expose(
    port: int,
    *,
    token: Optional[str] = None,
    expire: Optional[str] = None,
    daemon: bool = False,
    no_qr: bool = True,
    binary: Optional[str] = None,
    timeout: float = 30.0,
) -> Tunnel:
    """
    Expose ``port`` via the mekong binary.

    Returns a :class:`Tunnel` whose ``.url`` is the public HTTPS address.
    Use as a context manager or call ``.stop()`` when done.

    Parameters
    ----------
    port    : local port to expose
    token   : API token (falls back to MEKONG_TOKEN env or saved login)
    expire  : lifetime, e.g. "2h", "1d"
    daemon  : run tunnel in background (-d flag)
    no_qr   : suppress QR code output (default True in SDK mode)
    binary  : explicit path to mekong binary
    timeout : seconds to wait for the tunnel URL (default 30)
    """
    mekong_bin = binary or shutil.which("mekong")
    if not mekong_bin:
        # common install locations
        for candidate in [
            Path.home() / "bin" / "mekong",
            Path.home() / ".local" / "bin" / "mekong",
            Path("/usr/local/bin/mekong"),
            Path("/usr/bin/mekong"),
        ]:
            if candidate.exists():
                mekong_bin = str(candidate)
                break
    if not mekong_bin:
        raise FileNotFoundError(
            "mekong binary not found. "
            "Install from https://github.com/MuyleangIng/MekongTunnel/releases"
        )

    resolved_token = token or get_token()

    args = [mekong_bin, str(port)]
    if expire:
        args += ["--expire", expire]
    if resolved_token:
        args += ["--token", resolved_token]
    if daemon:
        args.append("-d")
    if no_qr:
        args.append("--no-qr")

    proc = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                             text=True, bufsize=1)

    url_found: list[str] = []
    error_ev = threading.Event()

    import re
    # Strip ANSI codes first, then extract clean URLs
    _ansi_re = re.compile(r'\x1b\[[0-9;]*[A-Za-z]')
    url_re   = re.compile(r"https?://[a-zA-Z0-9][a-zA-Z0-9\-./]*[a-zA-Z0-9]")

    def _reader(stream):
        for line in stream:
            if not url_found:
                clean = _ansi_re.sub("", line)
                m = url_re.search(clean)
                if m:
                    url_found.append(m.group(0))

    t_out = threading.Thread(target=_reader, args=(proc.stdout,), daemon=True)
    t_err = threading.Thread(target=_reader, args=(proc.stderr,), daemon=True)
    t_out.start()
    t_err.start()

    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if url_found:
            return Tunnel(url=url_found[0], port=port, _proc=proc)
        if proc.poll() is not None:
            raise RuntimeError(f"mekong exited with code {proc.returncode} before tunnel was ready")
        time.sleep(0.2)

    proc.terminate()
    raise TimeoutError(f"Timed out waiting for tunnel URL after {timeout}s")


# ── login() ──────────────────────────────────────────────────────────────────

def login(*, open_browser: bool = True) -> str:
    """
    Authenticate via the browser device flow.

    Opens ``angkorsearch.dev/cli-auth`` in the browser, polls until the user
    approves, then saves the token to ``~/.mekong/config.json``.

    Returns the raw API token string.
    """
    print("\nConnecting to angkorsearch.dev...", flush=True)

    try:
        raw = _api_post("/api/cli/device")
    except urllib.error.URLError as e:
        raise ConnectionError(f"Could not reach API: {e}") from e

    sess = _unwrap(raw)
    session_id  = sess["session_id"]
    login_url   = sess["login_url"]
    poll_interval = max(int(sess.get("poll_interval", 3)), 2)

    print(f"\n  Open this URL to log in:\n  \033[35m{login_url}\033[0m\n")

    if open_browser:
        try:
            webbrowser.open(login_url)
        except Exception:
            pass

    print("  Waiting for authorization", end="", flush=True)

    deadline = time.monotonic() + 5 * 60  # 5-minute window
    while time.monotonic() < deadline:
        time.sleep(poll_interval)
        try:
            pb = _api_get(f"/api/cli/device?session_id={session_id}")
        except Exception:
            print(".", end="", flush=True)
            continue

        poll = _unwrap(pb)
        status = poll.get("status")

        if status == "approved" and poll.get("token"):
            token = poll["token"]
            _save_config({"token": token})
            print("\n\n  \033[32m✔  Logged in!\033[0m\n")
            return token

        if status == "expired":
            raise RuntimeError("Session expired — call login() again")

        print(".", end="", flush=True)

    raise TimeoutError("Login timed out — call login() again")


# ── logout() ─────────────────────────────────────────────────────────────────

def logout() -> None:
    """Remove saved credentials from ~/.mekong/config.json."""
    try:
        _config_path().unlink()
        print("  Logged out.")
    except FileNotFoundError:
        print("  Already logged out.")


# ── whoami() ─────────────────────────────────────────────────────────────────

def whoami() -> Optional[dict]:
    """Return saved auth config dict or None if not logged in."""
    return _load_config()
